src/main/scala/com/fulcrumgenomics/bam/ComputeRnaSeqInsertSize.scala

```diff
@@ -0,0 +1,362 @@
+    val counters            = PairOrientation.values().toList.map { pairOrientation => (pairOrientation, new NumericCounter[Long]()) }.toMap
+    val filters             = ComputeRnaSeqInsertSize.filters(minimumMappingQuality=minimumMappingQuality, includeDuplicates=includeDuplicates)
+    val recordIterator      = new FilteringSamIterator(in.iterator(), new AggregateFilter(filters), false)
+
+    refFlatSource.foreach { gene => geneOverlapDetector.addLhs(gene, gene) }
+
+    recordIterator.foreach { rec =>
+      calculateInsertSize(rec=rec, overlapDetector=geneOverlapDetector) match {
+        case None             => Unit // ignore
+        case Some(insertSize) => counters(SamPairUtil.getPairOrientation(rec)).count(insertSize)
+      }
+      progress.record(rec)
+    }
+
+    refFlatSource.safelyClose()
+    in.safelyClose()
+
+    val metrics = PairOrientation.values().toList.map { pairOrientation =>
+      val counter = counters(pairOrientation)
+
+      // Create a metric *without* mean and standard deviation, as we will filter after
+      val median  = counter.median()
+      val mad     = counter.mad(m=median)
+      val metric = InsertSizeMetric(
+        pairOrientation           = pairOrientation,
+        read_pairs                = counter.total,
+        mean                      = 0,
+        median                    = median,
+        min                       = counter.map(_._2).min,
+        max                       = counter.map(_._2).max,
+        standard_deviation        = 0,
+        median_absolute_deviation = mad
+      )
+
+      // Remove outliers based on the mean absolute deviation and the # of allowable deviations
+      val maximumInsertSize = median + (deviations * mad)
+      val filteredCounter   = NumericCounter.from(counter.filter { case (value, _) => value <= maximumInsertSize })
+      val filteredMean      = filteredCounter.mean()
+
+      metric.copy(mean=filteredMean, standard_deviation=filteredCounter.stddev(m=filteredMean))
+    }
+
+    Metric.write(metrics=metrics, path=output)
+  }
+
+  /** Calculates the insert size in transcript space. */
+  private def calculateInsertSize(rec: SAMRecord, overlapDetector: OverlapDetector[Gene]): Option[Int] = {
+
+    val mateCigar        = SAMUtils.getMateCigar(rec)
+    require(mateCigar != null, s"Mate CIGAR (Tag MC) not found for $rec, consider using SetMateInformation or ClipOverlappingReads.")
+    val mateAlignmentEnd = CoordMath.getEnd(rec.getMateAlignmentStart, mateCigar.getReferenceLength)
+
+    // Get the overlapping genes
+    val leftMostBase     = Math.min(rec.getAlignmentStart, rec.getMateAlignmentStart)
+    val rightMostBase    = Math.max(rec.getAlignmentEnd, mateAlignmentEnd)
+    val recInterval      = new Interval(rec.getReferenceName, leftMostBase, rightMostBase)
+    val overlappingGenes = overlapDetector.getOverlaps(recInterval)
+    val firstGene        = overlappingGenes.headOption
```

Comment:
I don't think this helps you.  In the `if` below you can just `&& overlappingGenes.forall`, and on line 146 you have to unpack the option anway, so might as well be `overlappingGenes.head`.